pkg/iss/checkpoint.go

```diff
@@ -116,38 +122,44 @@ func (ct *checkpointTracker) ProcessAppSnapshotHash(snapshotHash []byte) *events
 	m := CheckpointMessage(ct.epoch, ct.seqNr, ct.appSnapshotHash)
 	walEvent.FollowUp(events.SendMessage(m, ct.membership))
 
-	// If the app snapshot was the last thing missing for the checkpoint to become stable,
-	// also produce the necessary events.
-	if ct.stable() {
-		walEvent.FollowUps(ct.announceStable().Slice())
+	// Apply pending Checkpoint messages
+	for s, m := range ct.pendingMessages {
+		walEvent.FollowUps(ct.applyMessage(m, s).Slice())
 	}
 
 	// Return resulting WALEvent (with the SendMessage event appended).
 	return (&events.EventList{}).PushBack(walEvent)
 }
 
-func (ct *checkpointTracker) applyMessage(chkpMsg *isspb.Checkpoint, source t.NodeID) *events.EventList {
+func (ct *checkpointTracker) applyMessage(msg *isspb.Checkpoint, source t.NodeID) *events.EventList {
 
 	// If checkpoint is already stable, ignore message.
 	if ct.stable() {
 		return &events.EventList{}
 	}
 
-	// TODO: Check signature of the sender.
-
-	// TODO: Distinguish messages by snapshot hash,
-	//       separately keeping the set of nodes from which a Checkpoint message has been received.
+	// Check snapshot hash
+	if ct.appSnapshotHash == nil {
+		// The message is received too early, put it aside
+		ct.pendingMessages[source] = msg
+		return &events.EventList{}
+	} else if !bytes.Equal(ct.appSnapshotHash, msg.AppSnapshotHash) {
+		// Snapshot hash mismatch
+		// TODO: Add diagnostic output
```

Comment:
@matejpavlovic how do I add diagnostic output here?

Reply:
In principle, this means that either the remote or the local replica is corrupt.

Reply:
The way I'd go about it would be to add a logger to the checkpoint tracker:
```go
type checkpointTracker struct {

        ...

	// Set of nodes from which any Checkpoint message has been received.
	// This is necessary for ignoring all but the first message a node sends, regardless of the snapshot hash.
	confirmations map[t.NodeID]struct{}

        logger logging.Logger // From the "github.com/filecoin-project/mir/pkg/logging" package
}
```

On creation of the checkpoint tracker in `getCheckpointTracker()`, pass a decorated logger to the `newCheckpointTracker()` function as an argument, following the same principle as when creating a new PBFT instance (https://github.com/filecoin-project/mir/blob/a64cd47b0ae4e046dbb57cf8cb50649fae009853/pkg/iss/iss.go#L637).
e.g.
```go
        iss.checkpoints[sn] = newCheckpointTracker(
            sn,
            logging.Decorate(iss.logger, "CT: ", "sn", sn),
        )
    
```

When the checkpoint tracker has a logger, you can simply print a message here through that logger.